### ISSUE TITLE
How can we use join with DAO objects?

### ISSUE THREAD

**Issue by msrd0** [JetBrains/Exposed#158]
Imagine two tables that reference each other, like this:

```kotlin
object Storages : LondIdTable()
{
    val user = reference("user", Users)
}
class Storage(id : EntityID<Long>) : LongEntity(id)
{
    companion object : LongEntityClass<Storage>(Storages)

    var user by User referencedOn Storages.user
}

object Users : LongIdTable()
{
     val name = text("name")
}
class User(id : EntityID<Long>) : LongEntity(id)
{
    companion object : LongEntityClass<User>(Users)

    var name by Users.name
}
```

Now, if we want to process all storages and retrieve their user, we could come up with this code:

```kotlin
transaction {
    Storage.all().forEach {
        println(it.user.name)
    }
}
```

With drivers that support multiple result sets like PostgreSQL, this works as expected. However, if the driver doesn't support multiple result sets like SQLite, we will only see the first storage's user name. _Why?_ **Because when querying the user of the storage, exposed starts a new query to the users table, which closes the result set of the query to the storages table.**

Fixing this would be as easy as retrieving both tables in one result set. The readme of Exposed shows ways to create join queries for non-DAO tables. But I cannot find any reference for join between DAO tables, and no method within `LongEntityClass` seems to handle this either.

### TL;DR

**Can we join DAO-tables for queries, or are there any plans to support this?**

**Comment by Tapac** [comment]
There are several ways to make what you want:
1. Load all Storages and then iterate over it. Just call toList() on SizedIterable to fetch it into memory. This will require 1 query to load Storages and N to load Users. (Not very effective).
```Kotlin
Storage.all().toList().forEach {
        println(it.user.name)
    }
```
2. Load both Storage and User in single query, load them to EntityCache and then work as usual. This approach requires one single query, but a bit more code.
```Kotlin
Storages.innerJoin(Users).map { 
    User.wrapRow(it)
    Storage.wrapRow(it)
}.forEach {
    println(it.user.name)
}
```
3. Load all Storages and then preload all User references to cache. 2 queries to rule them all.
```Kotlin
Storage.all().toList().apply {
   User.warmUpReferences(map{ it.id }, Storages.user)
}.forEach {
    println(it.user.name)
}
```
